[user triton] typing triton_kernel_wrap.py

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138230

Remove `# mypy: allow-untyped-defs` from triton_kernel_wrap.py, and fixed all the mypy errors.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

davidberard98 committed:
[user triton] typing triton_kernel_wrap.py

[ghstack-poisoned]

davidberard98 committed:
Update on "[user triton] typing triton_kernel_wrap.py"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

davidberard98 committed:
Update on "[user triton] typing triton_kernel_wrap.py"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

davidberard98 committed:
Update on "[user triton] typing triton_kernel_wrap.py"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

davidberard98 committed:
Update on "[user triton] typing triton_kernel_wrap.py"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

davidberard98 committed:
Update on "[user triton] typing triton_kernel_wrap.py"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

davidberard98 committed:
Update on "[user triton] typing triton_kernel_wrap.py"


Remove `# mypy: allow-untyped-defs` from triton_kernel_wrap.py, and fixed all the mypy errors.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

oulgen commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -67,21 +88,21 @@ def add_kernel(self, kernel) -> int:
     # Not every constant arg can be added to the graph. Use this side table
     # for constant args.
-    def add_constant_args(self, args) -> int:
+    def add_constant_args(self, args: Any) -> int:
```
instead of any use list[object]? any completely kills typechecking

oulgen commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -67,21 +88,21 @@ def add_kernel(self, kernel) -> int:
 
     # Returns the constant args
-    def get_constant_args(self, idx: int):
+    def get_constant_args(self, idx: int) -> Any:
```
ditto

oulgen commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -408,11 +433,19 @@ def mlir_to_functions(op) -> None:
 
 class MemoizeWithCycleCheck:
-    def __init__(self, fn):
+    fn: Any
```
callable something something

aorenste replied:
At least: `Callable[..., Any]`

oulgen commented on torch/utils/_triton.py:
```diff
@@ -110,3 +111,18 @@ def patch_triton_dtype_repr():
     triton.language.dtype.__repr__ = lambda self: dtype_to_string(self)
+
+
+if has_triton():
```
builder.py does exactly this too, maybe update it to use this definition

oulgen reviewed (approved):
Looks good, thanks for doing this! Just a note about Any, it is better to use object since Any completely kills typechecking

zou3519 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -546,12 +583,12 @@ def __init__(self) -> None:
-        grid,
+        kernel_idx: int,
+        constant_args_idx: int,
+        grid: List["TritonGridType"],
```
This is never a callable, right? I thought that Dynamo always applied the grid

davidberard98 replied:
I ran the tests and the only place where it's a callable is the test `triton_kernel_wrapper_functional_dense`. I'll leave it as is for now and if I get a chance I'll follow-up by changing the test and updating the types?

https://github.com/pytorch/pytorch/blob/a3172809a14f4927bea4aadc0376b5ae8c0d348a/test/inductor/test_triton_kernels.py#L97-L108

aorenste commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -67,21 +88,21 @@ def add_kernel(self, kernel) -> int:
     # Not every constant arg can be added to the graph. Use this side table
     # for constant args.
-    def add_constant_args(self, args) -> int:
+    def add_constant_args(self, args: Any) -> int:
```
Not your change but does `args` represent a single arg or multiple args? The name being plural confuses me... If it's actually multiple then it should probably take (and get below should return) a Sequence...

davidberard98 replied:
I think this is actually a single arg, so I'll leave this as `Any` (cc @oulgen).

oulgen replied:
Its a dict: https://fburl.com/code/7g7aurgi

so multiple items?

davidberard98 replied:
actually, this is not a single arg - updating the typing

zou3519 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -901,25 +948,44 @@ class TritonHOPifier:
-    def call_grid(self, grid, meta, tx):
+    def call_grid(
+        self,
+        grid: "Union[TritonGridCallableType, VariableTracker]",
+        meta: "Union[TritonMetaParamsType, VariableTracker]",
```
is there a way to specify... if one of these is a variable tracker then the other should be too?

aorenste replied:
Yes:
```
@overload
def call_grid(self, grid: TritonGridCallableType, meta: TritonMetaParamsType, tx: Optional["InstructionTranslator"]) -> Union[Tuple[Union[int, sympy.Expr, SymInt], ...], Tuple["Proxy", ...]]:
    ...
@overload
def call_grid(self, grid: VariableTracker, meta: VariableTracker, tx: Optional["InstructionTranslator"]) -> Union[Tuple[Union[int, sympy.Expr, SymInt], ...], Tuple["Proxy", ...]]:
    ...
def call_grid(
        self,
        grid: Union[TritonGridCallableType, VariableTracker],
        meta: Union[TritonMetaParamsType, VariableTracker],
        tx: Optional["InstructionTranslator"],
    ) -> Union[Tuple[Union[int, sympy.Expr, SymInt], ...], Tuple["Proxy", ...]]:
```
When you use `overload` the type-checker picks one of the overloaded APIs and ignores the non-overloaded one.  At runtime python ignores the overloaded APIs and uses the non-overloaded one.

davidberard98 replied:
I just marked this `type: ignore[no-untyped-def]` - see my comment in response to Richard

zou3519 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -901,25 +948,44 @@ class TritonHOPifier:
+        grids: List["TritonGridTupleType"],
+        combined_args: Dict[str, Any],
+        tx: Optional["InstructionTranslator"],
+    ) -> Optional["ConstantVariable"]:
```
In the non-Dynamo case this shouldn't be ConstantVariable

davidberard98 replied:
But this is the base class - `call_HOP` can _either_ return None, or ConstantVariable - hence `Optional[ConstantVariable]`. Does that sound right or am I missing something?

zou3519 replied:
Sorry, I didn't see the Optional part, you're right

zou3519 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1141,31 +1224,48 @@ def call_triton_kernel(self, variable, args, kwargs, tx):
-    def call_grid(self, grid, meta, tx):
+    def call_grid(
+        self,
+        grid: "Union[TritonGridCallableType, VariableTracker]",
+        meta: "Union[TritonMetaParamsType, VariableTracker]",
```
slightly weird because these can't be VariableTracker

davidberard98 replied:
@aorenste can I ask for your input on what to do here?

TritonHOPifier is the base class; and then DynamoTritonHOPifier and TracingTritonHOPifier are subclasses. Based on their implementations, I imagine their call_grid signatures could be (i.e. mirroring the snippet in another one of your comments):

```
TracingTritonHOPifier:  # call this "signature A"
    def call_grid(self, grid: TritonGridCallableType, meta: TritonMetaParamsType, tx: None) -> Tuple[Union[int, sympy.Expr, SymInt], ...]:

DynamoTritonHOPifier:    # call this "signature B"
    def call_grid(self, grid: VariableTracker, meta: VariableTracker, tx: InstructionTranslator) -> Tuple[Proxy, ...]:

TritonHOPifier:  # call this "signature C"
    def call_grid(self, grid: VariableTracker | TritonGridCallableType, meta: VariableTracker | TritonMetaParamsType, tx: Optional[InstructionTranslator]) -> Union[Tuple[Union[int, sympy.Expr, SymInt], ...], Tuple[Proxy, ...]]:
```

I can mark TritonHOPifier with "signature C", or with the `@overload` strategy you mentioned.

But then when I add types to (Tracing|Dynamo)TritonHOPifier, mypy requires the signatures of `call_grid` implementations on these classes to match (at least for the parameters; and in the case of using `@overload`, I need to copy the `@overload` declarations to the subclasses).

Any ideas on a cleaner way to add typing in this scenario?

davidberard98 replied:
okay, maybe I'll just not annotate the types on the base class and mark those as `type: ignore[no-untyped-def]`. That seems to work.

aorenste replied:
I think the problem here is that this API is just bad. I don't know the details other than what you said here but it sounds like TracingTritonHOPifier and DynamoTritonHOPifier's call_grid() are just unrelated functions that happen to have the same name? Do we actually need to define it on TritonHOPifier? Like, does anyone actually have a plain old TritonHOPifier and expect to call `call_grid` on it? Without knowing which specific subtype they have how do they know what to pass it?

I suspect that for now your plan of not annotating the base class (or use 'Any' just to indicate that someone has at least kind of thought about it and bailed) is probably reasonable for now.

davidberard98 replied:
Thanks for the input! I'll keep it as is then (with no annotations on the base class)

>  Like, does anyone actually have a plain old TritonHOPifier and expect to call call_grid on it?

No... 

> it sounds like TracingTritonHOPifier and DynamoTritonHOPifier's call_grid() are just unrelated functions that happen to have the same name

Well, logically they do the same thing, but one operates on actual grid objects, while the other one operates on dynamo VariableTrackers that represent grid objects...

aorenste replied:
> Thanks for the input! I'll keep it as is then (with no annotations on the base class)
> 
> > Like, does anyone actually have a plain old TritonHOPifier and expect to call call_grid on it?
> 
> No...

In which case (in the future) maybe the right thing is to just remove `call_grid` from the base class.

aorenste commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -901,25 +948,44 @@ class TritonHOPifier:
     """
 
-    def raise_unsupported(self, msg):
+    def raise_unsupported(self, msg: str) -> None:
```
```suggestion
    def raise_unsupported(self, msg: str) -> Never:
```

davidberard98 committed:
Update on "[user triton] typing triton_kernel_wrap.py"


Remove `# mypy: allow-untyped-defs` from triton_kernel_wrap.py, and fixed all the mypy errors.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

davidberard98 committed:
Update on "[user triton] typing triton_kernel_wrap.py"


Remove `# mypy: allow-untyped-defs` from triton_kernel_wrap.py, and fixed all the mypy errors.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot rebase -s

davidberard98 committed:
Update

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge

wdvr commented:
@pytorchmergebot revert -m 'Reverting this, as it started failing tests on main'

Error: AssertionError: test_distributed must not have initialized CUDA context on main process. - might be related to the imports?

[GH job link](https://github.com/pytorch/pytorch/actions/runs/11409829892/job/31751975567) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/2f61b69603756c1fcaef71b231e598df31e20f42)

wdvr commented:
@pytorchmergebot revert -m 'Reverting this, as it started failing tests on main' -c nosignal

davidberard98 commented:
@wdvr can you link which tests? is it the structured logging tests?

wdvr commented:
@davidberard98 The link is in this comment: https://github.com/pytorch/pytorch/pull/138230#issuecomment-2423354052

the test is: test/distributed/test_c10d_nccl.py

davidberard98 commented:
@pytorchbot rebase -b main

davidberard98 committed:
Update on "[user triton] typing triton_kernel_wrap.py"


Remove `# mypy: allow-untyped-defs` from triton_kernel_wrap.py, and fixed all the mypy errors.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

davidberard98 committed:
Update on "[user triton] typing triton_kernel_wrap.py"


Remove `# mypy: allow-untyped-defs` from triton_kernel_wrap.py, and fixed all the mypy errors.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

davidberard98 commented:
I removed the changes in `torch.utils._triton` (and `torch._dynamo.variables.builder`), which I think were causing the cuda initialization.

davidberard98 commented:
@pytorchbot merge
